dfttk/magnetism.py

```diff
@@ -220,3 +230,169 @@ def rearrange_sites_and_magmoms(config_dir):
+def clean_magmom_configs(
+    df: pd.DataFrame, spin_configs: pd.DataFrame, species: list[str]
+) -> pd.DataFrame:
+    """
+    Cleans the magnetic moment configurations by filtering species of interest,
+    assigning total sign columns, and checking for discontinuities in magmom.
+
+    Args:
+        df (pd.DataFrame): DataFrame containing the relaxed energy-volume data.
+        spin_configs (pd.DataFrame): DataFrame containing the original spin configurations from ATAT icamag.
+        species (list): List of species to filter by.
+
+    Returns:
+        pd.DataFrame: Cleaned DataFrame with consistent magnetic moment configurations.
+    """
+    # Keep only species of interest in df_copy and spin_configs_copy.
+    df_copy = df.copy()
+    df_copy["mag_data"] = df_copy.apply(filter_mag_data, species=species, axis=1)
+    spin_configs_copy = spin_configs.copy()
+    spin_configs_copy["mag_data"] = spin_configs_copy.apply(
+        filter_mag_data, species=species, axis=1
+    )
+
+    # Assign a tot_sign column to each mag_data.
+    df_copy["mag_data"] = df_copy.apply(assign_tot_sign, axis=1)
+    spin_configs_copy["mag_data"] = spin_configs_copy.apply(assign_tot_sign, axis=1)
+
+    unique_configs = spin_configs_copy["config"].unique()
+    match_list = []
+    index_list = []
+    unstable_initial_states = []
+    clean_df = pd.DataFrame()
+    for config in unique_configs:
+        # First, check for any discontinuities in magmom within a single configuration ev-curve.
+        config_df = df_copy[df_copy["config"] == config]
+        ref = spin_configs_copy[spin_configs_copy["config"] == config]
+        ref_tot_sign = ref["mag_data"].values[0]["tot_sign"]
+
+        count = 0
+        for index, row in config_df.iterrows():
+            current_tot_sign = row["mag_data"]["tot_sign"]
+            if count == 0:
+                match = current_tot_sign.equals(ref_tot_sign)
+            else:
+                previous_tot_sign = config_df.iloc[count - 1]["mag_data"]["tot_sign"]
+                match = current_tot_sign.equals(previous_tot_sign)
+
+            match_list.append(match)
+            index_list.append(index)
+            count += 1
+
+        # If all elements in match_list are true, there are no jumps, and it matches the original spin configuration.
+        if all(match_list):
+            multiplicity = ref["multiplicity"].values[0]
+            config_df.insert(1, "multiplicity", multiplicity)
+            clean_df = pd.concat([clean_df, config_df.loc[index_list]])
+
+        # If the first element is False but the rest are True, there is a jump from the original spin configuration only.
+        elif match_list[0] == False and all(match_list[1:]):
+            tot_sign = config_df.iloc[0]["mag_data"]["tot_sign"]
+            spin_up = (tot_sign[tot_sign == 1]).count()
+            spin_down = (tot_sign[tot_sign == -1]).count()
+            unstable_initial_states.append((config, spin_up, spin_down))
```

Comment:
This seems like a more complicated way to do this than to use `significant_magmom_change()` on each calculation, if I understand correctly.

Reply:
At the moment, I need to refer back to the original spin configs generated by icamag.